pkg/tracing/trace_context.go

```diff
@@ -0,0 +1,46 @@
+package tracing
+
+import (
+	"net/http"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/openzipkin/zipkin-go-opentracing"
+)
+
+type TraceContext interface {
+	Close() error
+	InitRequestSpan(r *http.Request) (FinishableSpan, error)
+}
+
+type FinishableSpan interface {
+	Finish()
+}
+
+type HTTPTraceContext struct {
+	Collector zipkintracer.Collector
+	Tracer    opentracing.Tracer
+	Span      opentracing.Span
+}
+
+type NoOpTraceContext struct{}
+type NoOpSpan struct{}
+
+func (c *HTTPTraceContext) Close() error {
+	return c.Collector.Close()
+}
+
+func (c *HTTPTraceContext) InitRequestSpan(r *http.Request) (FinishableSpan, error) {
+	span := c.Tracer.StartSpan("http-span") //TODO: improve naming
+
+	return span, c.Tracer.Inject(span.Context(), opentracing.TextMap, opentracing.HTTPHeadersCarrier(r.Header))
```

Comment:
As for the http-gw PR, even though it is less obvious, isn't there a way this could be achieved by wrapping the http.Client in a decorator?